src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,17 +105,131 @@ def pileupSizeTask(self):
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # usage of block names defined in this logic:
+            # https://github.com/dmwm/WMCore/pull/11807#pullrequestreview-1786778783
+            pname = doc['pileupName']
+            blockNames = self.rucioClient.getBlocksInContainer(pname)
+            if previousFraction < fraction:
+                # if it increasing
+                # use EVERY single block defined in the custom dataset plus blocks from the standard pileup
+                cname = doc.get('customName', '')
+                if cname:
+                    blockNames = self.rucioClient.getBlocksInContainer(cname) + blockNames
+            else:
+                # if it is decreasing
+                # instead of using random blocks from the standard pileup dataset (pileupName)
+                # we should use it from the custom dataset (customName) - if existent
+                cname = doc.get('customName', '')
+                if cname:
+                    blockNames = self.rucioClient.getBlocksInContainer(cname)
+
+            # get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+            portion = blockNames[:math.ceil(fraction * len(blockNames))]
+            self.logger.info("For pileup document %s take portion %s", doc, portion)
+
+            # create new container DID in Rucio for our custom name
+            self.rucioClient.createContainer(doc['customName'], scope=self.customRucioScope)
+
+            # we call attachDIDs Rucio wrapper API with our set of DIDs
+            newRules = []
+            for rse in doc['currentRSEs']:
+                self.rucioClient.attachDIDs(rse, doc['customName'], portion, scope=self.customRucioScope)
+
+                # create new rule for custom DID using pileup document rse
+                newRules += self.rucioClient.createReplicationRules(portion, rse)
+
+            self.logger.info("newRules %s", newRules)
+            # set expiration date (to be 24h) for already existing ruleIds from pileup document
+            for rid in doc['ruleIds']:
+                # set expiration date to be 24h ahead of right now
+                opts = {'lifetime': 24 * 60 * 60}
+                self.rucioClient.updateRule(rid, opts)
+
+            # update pileup document
+            if newRules:
+                # remove duplicate rules
+                newRules = set(newRules)
+                try:
+                    # update rules
+                    doc['ruleIds'] += newRules
+
+                    # update transition record of the pileup document, see logic:
+                    # https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5
+
+                    # we should always have transition record(s)
```

Comment:
I had a hard time following up on all these updates. So I would like to suggest just to reorganize it according to the description in point 7 of the gist). My suggestion would be to implement it like:
```
# update top level parameters
doc['ruleIds'] += newRules
doc['customName'] = cname
# update transition record
transition = doc['transition'][-1]
transition['containerFraction'] = fraction
transition['customDID'] = cname
transition['updateTime'] = gmtimeSeconds()
```

However, if you think these changes don't make the code more readable and sustainable, feel free to keep your implementation.

Reply:
ok done